File: fushi/test/pages/popup_pitch_merge_identical_test.js
// BUG-2122 behavior test: 同一个音调型被多本词典各渲染成一行。
//
// 用户报告（官网首页 demo 弹窗，查「ギター」）：音高区连出五行一模一样的
// `￣ギター [1]`，来源分别是五本音调词典。Yomitan 在 getGroupedPronunciations 里
// 把相同发音合并成一条、后面挂全部来源标签；popup.js 之前是一本词典一行。
//
// 本测试 EXECUTES 真实的 popup.js（vm + 极简假 DOM），驱动真实的
// `createPitchSection`，然后走产出的元素树数点 `.pitch-group` 行数与
// `.pitch-dict-label` 药丸。把 mergeIdenticalPitchGroups 或它的调用点撤掉，
// case 1 立刻变红。
//
// 覆盖：
//   1. 五本词典同为 [1]（去重关闭）→ 只剩 1 行，5 枚来源药丸，按首次出现顺序。
//   2. 音调型不同（[1] vs [0]）→ 不合并，仍是 2 行。
//   3. 位置部分重叠（[1,0] vs [1]）→ 判据是 payload 全等，故意不合并，仍是 2 行。
//   4. 去重打开 + 五本同为 [1] → 与改动前一致：1 行 1 枚药丸（默认外观零变化）。
//   5. 两本纯 IPA 词典给出完全相同的 transcriptions（去重打开）→ 合并成 1 行 2 枚药丸。
//
// Run: node fushi/test/pages/popup_pitch_merge_identical_test.js
// (also driven from popup_pitch_merge_identical_test.dart so it executes inside
//  `flutter test`).

const assert = require('assert');
const fs = require('fs');
const path = require('path');
const vm = require('vm');

const popupPath = path.resolve(__dirname, '../../assets/popup/popup.js');
const source = fs.readFileSync(popupPath, 'utf8');

function makeElement(tag) {
  return {
    tagName: (tag || 'div').toUpperCase(),
    className: '',
    id: '',
    textContent: '',
    innerHTML: '',
    nodeType: 1,
    style: {},
    dataset: {},
    children: [],
    childNodes: [],
    attributes: {},
    classList: {
      _set: new Set(),
      add(name) { this._set.add(name); },
      remove(name) { this._set.delete(name); },
      contains(name) { return this._set.has(name); },
    },
    appendChild(child) { this.children.push(child); this.childNodes.push(child); return child; },
    append(...nodes) { this.children.push(...nodes); this.childNodes.push(...nodes); },
    setAttribute(k, v) { this.attributes[k] = v; },
    removeAttribute(k) { delete this.attributes[k]; },
    addEventListener() {},
    querySelectorAll() { return []; },
    querySelector() { return null; },
    closest() { return null; },
  };
}

function makeTextNode(text) {
  return { nodeType: 3, textContent: String(text), children: [], childNodes: [] };
}

function makeSandbox() {
  const documentObj = {
    documentElement: { style: {}, classList: makeElement().classList },
    head: { appendChild() {} },
    body: makeElement('body'),
    getElementById() { return null; },
    querySelector() { return null; },
    querySelectorAll() { return []; },
    createElement(tag) { return makeElement(tag); },
    createTextNode(text) { return makeTextNode(text); },
    addEventListener() {},
  };

  const windowObj = {
    audioSources: [],
    needsAudio: false,
    lookupEntries: [],
    dictionaryStyles: {},
    // 官网 demo 与「关掉去重」的用户就是这一档；每个 case 各自覆写。
    deduplicatePitchAccents: false,
    flutter_inappwebview: { callHandler() { return Promise.resolve(false); } },
    getSelection() { return { toString() { return ''; } }; },
  };
  documentObj.defaultView = windowObj;

  const sandbox = {
    Node: { TEXT_NODE: 3, ELEMENT_NODE: 1 },
    Date, Math, URL, JSON, RegExp, Set, Map, Object, Array, console,
    performance: { now() { return 0; } },
    setTimeout, clearTimeout,
    DOMParser: class { parseFromString() { return { body: makeElement('body'), querySelectorAll() { return []; } }; } },
    document: documentObj,
    window: windowObj,
    getComputedStyle() { return {}; },
  };
  sandbox.globalThis = sandbox;
  return sandbox;
}

function loadPopup() {
  const sandbox = makeSandbox();
  vm.createContext(sandbox);
  const exported = source + `
    ;window.__test = {
      createPitchSection: createPitchSection,
    };
  `;
  vm.runInContext(exported, sandbox, { filename: 'popup.js' });
  return sandbox;
}

// 深度优先收集所有 className == cls 的元素节点。
function collectByClass(node, cls, acc) {
  acc = acc || [];
  if (!node) return acc;
  if (node.nodeType !== 3 && node.className === cls) acc.push(node);
  const kids = node.children || node.childNodes || [];
  for (const k of kids) collectByClass(k, cls, acc);
  return acc;
}

function collectText(node) {
  if (!node) return '';
  let out = node.nodeType === 3 ? (node.textContent || '') : '';
  if (typeof node.textContent === 'string' && node.nodeType !== 3 &&
      (!node.children || node.children.length === 0)) {
    out += node.textContent;
  }
  const kids = node.children || node.childNodes || [];
  for (const k of kids) out += collectText(k);
  return out;
}

function labelNames(section) {
  return collectByClass(section, 'pitch-dict-label').map(n => n.textContent);
}

function pitchGroupCount(section) {
  return collectByClass(section, 'pitch-group').length;
}

const FIVE_SAME = ['词典14', '词典13', '词典15', '词典16', '词典17'].map(
  name => ({ dictionary: name, pitchPositions: [1], patterns: [], transcriptions: [] }));

(function run() {
  // Case 1: 用户报告的原样输入 —— 五本词典同为 [1]，去重关闭。
  {
    const sb = loadPopup();
    sb.window.deduplicatePitchAccents = false;
    const section = sb.window.__test.createPitchSection(FIVE_SAME, 'ギター');
    assert.ok(section, 'five identical pitch dicts must still render a pitch section');
    assert.strictEqual(pitchGroupCount(section), 1,
      'five dictionaries agreeing on [1] must collapse into ONE .pitch-group row; got '
        + pitchGroupCount(section));
    assert.deepStrictEqual(labelNames(section),
      ['词典14', '词典13', '词典15', '词典16', '词典17'],
      'the merged row must carry every source label, in first-appearance order');
    const text = collectText(section);
    const occurrences = text.split('[1]').length - 1;
    assert.strictEqual(occurrences, 1,
      'the accent [1] must be drawn exactly once after merging; got ' + occurrences
        + ' in ' + JSON.stringify(text));
  }

  // Case 2: 音调型不同 —— 绝不合并。
  {
    const sb = loadPopup();
    sb.window.deduplicatePitchAccents = false;
    const section = sb.window.__test.createPitchSection([
      { dictionary: 'A', pitchPositions: [1], patterns: [], transcriptions: [] },
      { dictionary: 'B', pitchPositions: [0], patterns: [], transcriptions: [] },
    ], 'ねこ');
    assert.strictEqual(pitchGroupCount(section), 2,
      'dictionaries disagreeing on the accent must stay on separate rows');
    assert.deepStrictEqual(labelNames(section), ['A', 'B']);
  }

  // Case 3: 位置部分重叠 —— 判据是 payload 全等，故意不合并（宁可少合）。
  {
    const sb = loadPopup();
    sb.window.deduplicatePitchAccents = false;
    const section = sb.window.__test.createPitchSection([
      { dictionary: 'A', pitchPositions: [1, 0], patterns: [], transcriptions: [] },
      { dictionary: 'B', pitchPositions: [1], patterns: [], transcriptions: [] },
    ], 'ねこ');
    assert.strictEqual(pitchGroupCount(section), 2,
      'partially overlapping position sets must NOT be merged (payload equality is the rule)');
  }

  // Case 4: 去重打开（app 默认）—— 行为与改动前逐字一致：1 行 1 枚药丸。
  {
    const sb = loadPopup();
    sb.window.deduplicatePitchAccents = true;
    const section = sb.window.__test.createPitchSection(FIVE_SAME, 'ギター');
    assert.strictEqual(pitchGroupCount(section), 1,
      'dedup ON must still yield exactly one row');
    assert.deepStrictEqual(labelNames(section), ['词典14'],
      'dedup ON keeps only the first dictionary — the default look must not change');
  }

  // Case 5: 两本纯 IPA 词典给出完全相同的 transcriptions → 合并。
  {
    const sb = loadPopup();
    sb.window.deduplicatePitchAccents = true;
    const section = sb.window.__test.createPitchSection([
      { dictionary: 'IPA-1', pitchPositions: [], patterns: [], transcriptions: ['neꜜko'] },
      { dictionary: 'IPA-2', pitchPositions: [], patterns: [], transcriptions: ['neꜜko'] },
    ], 'ねこ');
    assert.strictEqual(pitchGroupCount(section), 1,
      'two IPA dicts with identical transcriptions must merge into one row');
    assert.deepStrictEqual(labelNames(section), ['IPA-1', 'IPA-2']);
    const text = collectText(section);
    assert.strictEqual(text.split('[neꜜko]').length - 1, 1,
      'the shared transcription must be printed once; got ' + JSON.stringify(text));
  }

  console.log('popup_pitch_merge_identical_test.js: all assertions passed');
})();

File: fushi/test/pages/popup_pitch_merge_identical_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// BUG-2122：音高区里同一个音调型被多本词典各占一行。
///
/// 用户报告（官网首页 demo 弹窗查「ギター」）：音高区连出五行一模一样的
/// `￣ギター [1]`，只有前面的来源药丸不同。Yomitan 的 `getGroupedPronunciations`
/// 把相同发音合并成一条、后面挂全部来源；popup.js 之前是一本词典一行。
///
/// 修法：渲染前加一道纯变换 `mergeIdenticalPitchGroups` —— 整份 payload 全等
/// （pitchPositions / patterns / transcriptions 三者）的词典合并成一行，
/// `dictionaries` 带上全部来源名。它跑在 `deduplicatePitchAccents` 分支**之后**：
/// 去重打开（app 默认）时各存活组的位置互斥、合并对位置组恒为 no-op，默认外观
/// 一字不变；去重关闭（官网 demo 就是这一档）时才塌行。
///
/// 两层守护：
/// ① 行为级——用 Node 真执行 popup.js 的 `createPitchSection`，断言五本同型词典
///    塌成 1 行 5 枚药丸、音调型不同的不合并、去重打开时行为与改动前逐字一致。
///    无 node 时 skip。
/// ② 源码级——静态扫描**三份镜像副本**（app / 扩展 assets / 扩展 tools），保证合并
///    helper、调用点、多药丸渲染和 CSS 换行在位；即便无 node 也能守住回归，且三份
///    副本不会漂移。
void main() {
  const Map<String, String> jsMirrors = <String, String>{
    'app popup': 'assets/popup/popup.js',
    'extension vendor (assets)': 'assets/browser_extension/vendor/popup.js',
    'extension vendor (tools)': '../tools/browser-extension/vendor/popup.js',
  };
  const Map<String, String> cssMirrors = <String, String>{
    'app popup': 'assets/popup/popup.css',
    'extension vendor (assets)': 'assets/browser_extension/vendor/popup.css',
    'extension vendor (tools)': '../tools/browser-extension/vendor/popup.css',
  };

  test(
    'identical pitch accents from several dictionaries merge into one row '
    '(executes createPitchSection via node)',
    () async {
      final String? nodeExe = _resolveNode();
      if (nodeExe == null) {
        markTestSkipped(
            'node not found on PATH; skipping JS behavior execution');
        return;
      }

      final File jsTest =
          File('test/pages/popup_pitch_merge_identical_test.js');
      expect(
        jsTest.existsSync(),
        isTrue,
        reason: 'behavior harness ${jsTest.path} must exist',
      );

      final ProcessResult result = await Process.run(
        nodeExe,
        <String>[jsTest.path],
        workingDirectory: Directory.current.path,
      );

      expect(
        result.exitCode,
        0,
        reason: 'popup pitch merge JS behavior test failed.\n'
            'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
      );
      expect(
        result.stdout.toString(),
        contains('all assertions passed'),
        reason: 'behavior harness must reach its success marker',
      );
    },
  );

  jsMirrors.forEach((String name, String relPath) {
    test('[$name] popup.js merges identical pitch groups before rendering', () {
      final File file = File(relPath);
      expect(file.existsSync(), isTrue, reason: '$relPath must exist');
      final String js = file.readAsStringSync();

      // 合并 helper 必须存在，且判据是整份 payload（三者全等），不是只看位置。
      final int helper =
          js.indexOf('function mergeIdenticalPitchGroups(groups)');
      expect(
        helper,
        greaterThanOrEqualTo(0),
        reason: 'the pitch merge helper must exist',
      );
      for (final String field in <String>[
        'group.pitchPositions || []',
        'group.patterns || []',
        'group.transcriptions || []',
      ]) {
        expect(
          js.indexOf(field, helper),
          greaterThan(helper),
          reason: 'the merge key must cover $field — merging on a partial '
              'payload would fuse dictionaries that disagree',
        );
      }

      // 非恒真：helper 存在但没人调用等于没改。调用点必须在 createPitchSection 内，
      // 且在去重分支之后（去重先跑，合并是最后一道变换）。
      final int section =
          js.indexOf('function createPitchSection(pitches, reading)');
      expect(section, greaterThanOrEqualTo(0));
      final int dedup =
          js.indexOf('if (window.deduplicatePitchAccents)', section);
      expect(dedup, greaterThan(section));
      final int call =
          js.indexOf('mergeIdenticalPitchGroups(groups).forEach(', section);
      expect(
        call,
        greaterThan(dedup),
        reason: 'createPitchSection must run the merge AFTER the dedup branch, '
            'so dedup ON keeps the pre-change default look',
      );

      // 去重分支不得再直接 append，必须先攒进 groups 交给合并。
      expect(
        js.substring(dedup, call),
        isNot(contains('pitchContainer.appendChild(createPitchGroup(')),
        reason: 'the dedup branch must collect groups instead of appending '
            'rows directly, or merging never sees them',
      );

      // createPitchGroup 每本来源渲染一枚药丸。
      final int group =
          js.indexOf('function createPitchGroup(pitchData, reading)');
      expect(group, greaterThanOrEqualTo(0));
      expect(
        js.indexOf(
            'const dictionaries = pitchData.dictionaries || [pitchData.dictionary];',
            group),
        greaterThan(group),
        reason: 'createPitchGroup must accept a dictionary LIST (single-source '
            'groups degrade to a one-element list)',
      );
      final int labelLoop =
          js.indexOf('dictionaries.forEach((dictionary) => {', group);
      expect(
        labelLoop,
        greaterThan(group),
        reason: 'every merged source must get its own .pitch-dict-label pill',
      );
      expect(
        js.indexOf("className: 'pitch-dict-label', textContent: dictionary",
            labelLoop),
        greaterThan(labelLoop),
        reason: 'the pill loop must render the per-dictionary label',
      );
    });
  });

  cssMirrors.forEach((String name, String relPath) {
    test('[$name] popup.css lets a merged pitch row wrap its source pills', () {
      final File file = File(relPath);
      expect(file.existsSync(), isTrue, reason: '$relPath must exist');
      final String css = file.readAsStringSync();

      final int rule = css.indexOf('.pitch-group {');
      expect(rule, greaterThanOrEqualTo(0),
          reason: '.pitch-group rule must exist');
      final int end = css.indexOf('}', rule);
      expect(end, greaterThan(rule));
      final String body = css.substring(rule, end);
      expect(
        body,
        contains('flex-wrap: wrap'),
        reason: 'a merged row can carry several source pills; without wrapping '
            'they push the reading past the popup edge',
      );
      // 换行只在 flex 容器上有意义——顺带钉住这条规则仍是 flex。
      expect(body, contains('display: flex'));
    });
  });
}

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) {
        return name;
      }
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}
